dialer.go

```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
-	d.muNumLivePending.Unlock()
-	return numLivePending
+func (d *dialer) Dial() (net.Conn, error) {
+	return d.DialContext(context.Background())
 }
 
-func (d *dialer) getOrCreateSession(ctx context.Context, dial DialFN) (sessionIntf, error) {
-	newSession := func(cap int) {
-		d.muNumLivePending.Lock()
-		if d.numLive+d.numPending >= cap {
-			d.muNumLivePending.Unlock()
-			return
-		}
-		d.numPending++
-		d.muNumLivePending.Unlock()
-		go func() {
-			s, err := d.startSession(dial)
-			d.muNumLivePending.Lock()
-			d.numPending--
-			if err != nil {
-				d.muNumLivePending.Unlock()
-				return
-			}
-			d.numLive++
-			d.muNumLivePending.Unlock()
-			d.liveSessions <- s
-		}()
+func (d *dialer) DialContext(ctx context.Context) (net.Conn, error) {
+	s, err := d.getSession(ctx)
+	if err != nil {
+		return nil, err
+	}
+	c := s.createStream(ctx)
+	select {
+	case d.liveSessions <- s:
+	default:
+		log.Debugf("Maximum live sessions reached to %v", d.name)
 	}
+
+	return c, nil
+}
+
+func (d *dialer) getSession(ctx context.Context) (sessionIntf, error) {
+	start := time.Now()
 	for {
 		select {
 		case s := <-d.liveSessions:
-			if s.AllowNewStream(d.maxStreamsPerConn, d.idleInterval) {
-				return s, nil
+			if s.isClosed() {
+				// Closed sessions will trigger the creation of new ones, so keep waiting for a new session.
+				continue
 			}
-			d.muNumLivePending.Lock()
-			d.numLive--
-			d.muNumLivePending.Unlock()
-			s.MarkDefunct()
-			newSession(minLiveConns)
-		case <-time.After(d.redialSessionInterval):
-			newSession(d.maxLiveConns)
+
+			if !s.allowNewStream(d.maxStreamsPerConn) {
+				log.Debugf("Maximum streams reached for session to %v", d.name)
+				// The default number of streams per session is 65535, so this is unlikely to be reached.
+				select {
+				case d.pendingSessions <- newPendingSession(d.name):
+				default:
+				}
+				continue
+			}
+
+			return s, nil
+
 		case <-ctx.Done():
-			return nil, errors.New("No session available")
+			elapsed := time.Since(start).Seconds()
+			err := fmt.Errorf("no session available after %f seconds to %v", elapsed, d.name)
+			return nil, err
 		}
 	}
 }
 
-func (d *dialer) returnSession(s sessionIntf) {
-	addBack := true
-	d.muNumLivePending.Lock()
-	if d.numLive > minLiveConns {
-		d.numLive--
-		addBack = false
-	}
-	d.muNumLivePending.Unlock()
-	if addBack {
-		d.liveSessions <- s
-	} else {
-		s.MarkDefunct()
-	}
-}
-
 func (d *dialer) EMARTT() time.Duration {
 	return d.emaRTT.GetDuration()
 }
 
-func (d *dialer) BoundTo(dial DialFN) BoundDialer {
-	return &boundDialer{d, dial}
-}
-
-func (d *dialer) startSession(dial DialFN) (*session, error) {
-	conn, err := dial()
+func (d *dialer) startSession(rs *pendingSession) (*session, error) {
+	lc := d.lifecyle.OnTCPStart()
+	conn, err := d.dial(rs.dialTimeout)
 	if err != nil {
+		lc.OnTCPConnectionError(err)
 		return nil, err
 	}
+	lc.OnTCPEstablished(conn)
 
 	cs, err := newCryptoSpec(d.cipherCode)
```

Comment:
This is not a new thing, but I noticed that if we get an error here or later, we don't close the conn. I think we should. It's likely not a problem in practice because we should never actually get errors here.

Reply:
nice catch will fix!

Reply:
I ended up just moving those calls up to before we create the TCP connection.